### PR TITLE
Use frozen NumPy array for ImageDecoder backend test if no Pillow

### PR DESCRIPTION
**Pull request by jcwchen**
### Description

- Use frozen NumPy array for ImageDecoder backend test's input/output if there is no installed pillow
- Test with and without pillow in CIs when running test generation

### Motivation and Context

Similar to https://github.com/onnx/onnx/pull/5498. Hopefully this PR can entirely fix https://github.com/onnx/onnx/issues/5491.

Ideally test generation should be as simple as possible without other new dependency especially that dependency was not used in other places (reference will only use pillow when needed). Let's merge this PR first and later I will have another PR to freeze input/output as a pure NumPy array and simply keep generate_checkerboard/generate_test_data for future reference.

### CODE REVIEW

**Review by jcwchen — COMMENTED** [review]
However, these test data are actually image so the array is quite big. They look very unorganized in the introduced Python file: image_decoder_data.py with 35637 lines... If anyone has better idea to handle them, feel free to let me know. Thanks!

**Comment by justinchuby** [comment]
Can we use smaller images? Make just a few pixels?

**Review by justinchuby — APPROVED** [review]

**Comment by lutzroeder** [comment]
@jcwchen validated this branch fixes the underlying issue. `collect_snippets` runs without requiring `cv2`.

File: onnx/backend/test/case/node/image_decoder.py
Comment (adityagoel4512) [inline onnx/backend/test/case/node/image_decoder.py:47]: Perhaps something like 
```suggestion
            "opencv-python must be installed to generate test data for the ImageDecoder operator"
```

Comment (jcwchen) [reply onnx/backend/test/case/node/image_decoder.py]: Good catch. Updated. Thanks!

**Review by jcwchen — COMMENTED** [review]
> Can we use smaller images? Make just a few pixels?

Good question. Now by default generate_test_data is using 40x40, which is sort of in the middle, but I don't have proper number in mind. Ideally the size of image should be large enough for runtime to test. @jantonguirao do you think it's a good idea to shrink the test image a bit?

**Comment by jantonguirao** [comment]
@jcwchen We could reduce the image size to something like 32x32, but I wouldn't go much smaller than that.

**Review by jcwchen — COMMENTED** [review]
> @jcwchen We could reduce the image size to something like 32x32, but I wouldn't go much smaller than that.

Thank you for the suggestion! I have used 32x32 instead of 40x40 and now the used lines in that Python file was reduced from  35637 lines to 23,604 lines.

File: onnx/backend/test/case/node/image_decoder.py
Comment (jcwchen) [inline onnx/backend/test/case/node/image_decoder.py:50]: Finally I decide to change the behavior as:
1. If the env has installed opencv-python, still use opencv-python to generate input/output.
2. If the env does not have installed opencv-python, use frozen data from image_decoder_data.py instead. I only comment this behavior in the code since it's more like implementation detail. By contrast, printing out the message will be quite redundant when running test generation.

I also verifies these 2 behaviors in the CIs.

**Comment by justinchuby** [comment]
We need this to be in the next weekly to unblock some tests. cc @gramalingam

**Comment by gramalingam** [comment]
A couple of questions:
* Is it worth saving the frozen data as a binary file? Is there any advantage to saving it as a python file?
* Do the tests only check if it works without open-cv installed? Is it worth checking that the data is the same with and without open-cv?

**Comment by jcwchen** [comment]
> Is it worth saving the frozen data as a binary file? Is there any advantage to saving it as a python file?

Good question. Initially I also considered to directly use the original intput0.pb and output0.pb, but I just feel having another reference to get the actual data would be safer in case someone accidentally update input/output pb files somehow. Or we can use another binary files to convert them as input/output pb, but currently we don't have any existing node test doing that. So I still chose using NumPy to create these data, just like what other node tests do, to keep it simple and more readable. Still, I am open to either way. Please let me know if you think it's better to use binary files.


> Do the tests only check if it works without open-cv installed? Is it worth checking that the data is the same with and without open-cv?

I have added tests in CI pipelines for node test generation with/without opencv as follows:

https://github.com/onnx/onnx/blob/9762bcfef2236554f8afa74dfdb980a41252129d/.azure-pipelines/Linux-CI.yml#L95-L118

- With opencv, backend test will use opencv to generate intput/output for ImageDecoder
- Without opencv, backend test will use frozen NumPy array from image_decoder_data.py.

**Comment by xadupre** [comment]
The proposal seems rather complex to me. Why not using pillow instead of opencv?

**Comment by jcwchen** [comment]
As @justinchuby suggested, I turned off black formatter for image_decoder_data.py and rearranged those large NumPy arrays. Now the code lines was decreased from 23,604 to 5,134, which is now previewable in this PR. @onnx/sig-operators-approvers PTAL. Thanks!

**Review by justinchuby — APPROVED** [review]

**Comment by justinchuby** [comment]
Would be nice to wait for #5555 because there are slight changes in values

**Comment by jcwchen** [comment]
I have updated _image_decoder_data.py accordingly with the new data from https://github.com/onnx/onnx/pull/5555 (created from Pillow). Also CI pipeline will test with/without pillow with node test generation.

If no one has other concern for this PR, I will forward this PR by the end of today. Thank you all for the reviews!

**Pull request merged**
